File: 3372-maximize-the-number-of-target-nodes-after-connecting-trees-i/3372-maximize-the-number-of-target-nodes-after-connecting-trees-i.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    // DFS 遞迴：計算從 node 出發，在距離 k 內可以 reach 到幾個節點（包含自己）
    int dfs(int node, int parent, vector<vector<int>>& children, int k) {
        if (k < 0) return 0;  // 超出距離範圍

        int res = 1;  // 包含自己
        for (int child : children[node]) {
            if (child == parent) continue;  // 避免走回來
            res += dfs(child, node, children, k - 1);  // 繼續往下遞迴
        }
        return res;
    }

    // 建立圖並計算每個節點在距離 <= k 內可以 reach 到的節點數量
    vector<int> build(vector<vector<int>>& edges, int k) {
        int n = edges.size() + 1;
        vector<vector<int>> children(n);  // 鄰接表
        for (auto& e : edges) {
            int u = e[0], v = e[1];
            children[u].push_back(v);
            children[v].push_back(u);
        }

        vector<int> res(n);
        for (int i = 0; i < n; ++i) {
            res[i] = dfs(i, -1, children, k);
        }
        return res;
    }

    // 主函式：傳回每個第一棵樹的節點最大 target 數量
    vector<int> maxTargetNodes(vector<vector<int>>& edges1, vector<vector<int>>& edges2, int k) {
        int n = edges1.size() + 1;

        // 計算第一棵樹中每個節點在距離 <= k 可以 reach 到的節點數
        vector<int> count1 = build(edges1, k);

        // 計算第二棵樹中每個節點在距離 <= k - 1 可以 reach 到的節點數
        vector<int> count2 = build(edges2, k - 1);

        // 找出第二棵樹中最大可達節點數
        int maxCount2 = *max_element(count2.begin(), count2.end());

        // 最終結果：第一棵樹中每個節點加上最大可能連接的節點數
        vector<int> res(n);
        for (int i = 0; i < n; ++i) {
            res[i] = count1[i] + maxCount2;
        }
        return res;
    }
};

File: 3372-maximize-the-number-of-target-nodes-after-connecting-trees-i/3372-maximize-the-number-of-target-nodes-after-connecting-trees-i.py
class Solution:
    def maxTargetNodes(
        self, edges1: List[List[int]], edges2: List[List[int]], k: int
    ) -> List[int]:
        
        # DFS：從 node 出發，往子節點遞迴，統計在距離 k 內可以訪問多少個節點（包含自己）
        def dfs(
            node: int, parent: int, children: List[List[int]], k: int
        ) -> int:
            if k < 0:
                return 0  # 距離超過 k，不能再往下遞迴
            res = 1  # 自己本身也算一個 target node
            for child in children[node]:
                if child == parent:
                    continue  # 避免走回父節點
                res += dfs(child, node, children, k - 1)  # 對每個子節點遞迴，距離減 1
            return res

        # 建樹與計算每個節點能 reach 到幾個節點（距離 <= k）
        def build(edges: List[List[int]], k: int) -> List[int]:
            n = len(edges) + 1  # 節點數等於邊數 + 1
            children = [[] for _ in range(n)]  # 建立鄰接表
            for u, v in edges:
                children[u].append(v)
                children[v].append(u)
            res = [0] * n
            for i in range(n):
                res[i] = dfs(i, -1, children, k)  # 對每個節點做 DFS 統計可達節點數
            return res

        n = len(edges1) + 1  # 第一棵樹的節點數
        count1 = build(edges1, k)        # 計算每個節點在第一棵樹中距離 ≤ k 可達節點數
        count2 = build(edges2, k - 1)    # 第二棵樹距離上限是 k-1，因為會多一條邊連接到第一棵樹
        maxCount2 = max(count2)          # 第二棵樹中，任一節點作為連接點時最多可達節點數

        # 對第一棵樹中的每個節點 i，取其可達節點數 + 第二棵樹中最大可達數，即最大可能 target 數
        res = [count1[i] + maxCount2 for i in range(n)]
        return res
